docs/js-guide/migrating.md

```diff
@@ -142,3 +143,68 @@ See [our approach to inheritance](https://github.com/dimagi/js-guide/blob/master
 [Sample pull request](https://github.com/dimagi/commcare-hq/pull/19938)
 
 Code that actually manipulates the prototype needs more thought.
+
+## Upgrading Select2
+
+HQ uses two different versions of select2. We're in the process of upgrading usage of the old version,
+because in principle it's crazy to have two versions but also because practically, having multiple versions
+makes it harder to bundle and share code across pages effectively.
+
+|                | New Select2                        | Old Select2                       |
+| -------------- | ---------------------------------- | --------------------------------- |
+| Version        |  4.0.0                             | 3.5.2                             |
+| Docs           | https://select2.org/               | http://select2.github.io/select2/ |
+| JS module      | `select2/dist/js/select2.full.min` | `select2-3.5.2-legacy/select2`    |
+| View decorator | `@use_select2_v4`                  | `@use_select2`                    |
+
+This is a fairly complicated migration both because the old and new version differ in multiple significant ways and because of how code is shared in HQ.
+
+### Figuring out what to migrate
+
+Migrating one select2 widget typically leads to migrating a number of other dependent ones. There are a couple of ways to pick a starting point:
+
+- Pick a view that uses the `use_select2` decorator
+- Pick a javascript module that directly depends on `select2-3.5.2-legacy/select2`
+- Pick a page that includes a script tag for `select2-3.5.2-legacy/select2.js` (rare, because most pages use the `@use_select2` decorator rather than including a script tag)
+- Browse the [requirejs bundle configuration](https://www.commcarehq.org/static/build.txt) to find modules that indirectly depend on `select2-3.5.2-legacy/select2`
+- Javascript modules that don't yet declare dependencies but that call the `.select2()` function
+- Pages that use a select2-dependent knockout binding like `select2` or `questionsSelect`
+
+### Migrating
+
+The two versions of select2 have several types of differences.
+
+- Javascript: the new API has numerous differences. There are several select2-heavy javascript modules that have both an old and new version, so it can be helpful to look for javascript filenames ending `v3.js` and the corresponding files ending in `v4.js` to see how the API has changed. Changes frequently needed in HQ code:
+   - `formatResult` and `formatSelection` have been renamed to `templateResult` and `templateSelection`
+   - The `results` option for ajax is renamed to `processResults`
+   - `placeholder` is now required if `allowClear` is true (set it to `' '` if there isn't placeholder text)
+   - `formatNoMatches` text for when there are no results is now part of the `language` option, the property `noResults`
+   - The `data` option for ajax now takes a `params` parameter with a `term` property instead of having `term` passed as a parameter.
```

Comment:
These two might be easier to understand by example, such as:
> instead of `data: function (term) {...}` use `data: function ({'term': term})`

Reply:
updated